pkg/subsystem/linux/parents.go

```diff
@@ -0,0 +1,162 @@
+		inInput[item] = true
+	}
+	matrix.NonEmptyPairs(func(anyA, anyB interface{}, count int) {
+		a, b := anyA.(*entity.Subsystem), anyB.(*entity.Subsystem)
+		if !inInput[a] || !inInput[b] {
+			return
+		}
+		// Demand that >= 50% paths are related.
+		if 2*count/matrix.Count(a) >= 1 && matrix.Count(a) < matrix.Count(b) {
+			a.Parents = append(a.Parents, b)
+		}
+	})
+	// Just in case.
+	if loopsExist(list) {
+		return fmt.Errorf("there are loops in the parents relation")
+	}
+	transitiveReduction(list)
+	return nil
+}
+
+// DropEmptySubsystems removes subsystems for which we have found no matches in the filesystem tree.
+func DropEmptySubsystems(matrix *match.CoincidenceMatrix, list []*entity.Subsystem) []*entity.Subsystem {
+	newList := []*entity.Subsystem{}
+	for _, item := range list {
+		if matrix.Count(item) > 0 {
+			newList = append(newList, item)
+		}
+	}
+	return newList
+}
+
+// DropDuplicateSubsystems makes sure there are no duplicate subsystems.
+// First, if subsystems A and B 100% overlap, we prefer the one that's alphabetically first.
+// Second, if subsystem A is fully enclosed in subsystem B and constitutes more than 75% of B,
+// we drop A, since it brings little value.
+func DropDuplicateSubsystems(matrix *match.CoincidenceMatrix, list []*entity.Subsystem) []*entity.Subsystem {
+	drop := map[*entity.Subsystem]struct{}{}
+	nameIsBetter := func(first, second string) bool {
+		// Let's prefer shorter names first.
+		if len(first) < len(second) {
+			return true
+		}
+		return first < second
+	}
+	matrix.NonEmptyPairs(func(anyA, anyB interface{}, count int) {
+		a, b := anyA.(*entity.Subsystem), anyB.(*entity.Subsystem)
+		// Only consider cases when A is fully enclosed in B, i.e. M[A][B] == M[A][A].
+		if count != matrix.Count(a) {
+			return
+		}
+		// If A and B 100% coincide, eliminate A and keep B if A > B.
+		if count == matrix.Count(b) {
+			if nameIsBetter(a.Name, b.Name) {
+				return
+			}
+			drop[a] = struct{}{}
+			return
+		}
+		// If A constitutes > 75% of B, drop A.
+		if 4*matrix.Count(a)/matrix.Count(b) >= 3 {
+			drop[a] = struct{}{}
+		}
+	})
+	newList := []*entity.Subsystem{}
+	for _, item := range list {
+		if _, exists := drop[item]; !exists {
+			newList = append(newList, item)
+		}
+	}
+	return newList
+}
+
+// The algorithm runs in O(E * (E + V)).
+// We expect that E is quite low here, so it should be fine.
+func transitiveReduction(list []*entity.Subsystem) {
+	for _, s := range list {
+		removeParents := map[*entity.Subsystem]bool{}
+		for _, p := range s.Parents {
+			for otherP := range p.ReachableParents() {
+				removeParents[otherP] = true
+			}
+		}
+		newParents := []*entity.Subsystem{}
+		for _, p := range s.Parents {
+			if !removeParents[p] {
+				newParents = append(newParents, p)
+			}
+		}
+		s.Parents = newParents
+	}
+}
+
+// loopsExist is a helper method that verifies that the resulting graph has no loops.
+func loopsExist(list []*entity.Subsystem) bool {
+	type graphNode struct {
```

Comment:
Such a check would give us false positives. E.g. if A has children B and C, which are both parents of D. This is not the kind of loop we want to avoid, we only want to prevent circular paths in this directed graph.

Reply:
Right, these subsystems with multiple parents always confuse me.
Then I hope the following should do:
Then `if p == subsystem { panic("a loop") }` in `ReachableParents` and calling `ReachableParents` after adding each new parent in `setParents`:
```
    a.Parents = append(a.Parents, b)
    a.ReachableParents() // ensure we are not adding a loop
```